### PR TITLE
Move theme management responsability to react-admin

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-ui-materialui/src/layout/Theme/useSetTheme.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { useContext } from 'react';
+import { DeprecatedThemeOptions } from '@mui/material';
+import { ThemeContext, ThemeSetter } from './ThemeContext';
+
+export const useSetTheme = (
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Theme/useSetTheme.ts:5]: Please add jsDoc for usage

File: packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import {
+    createTheme,
+    ThemeProvider as MuiThemeProvider,
+    Theme,
+} from '@mui/material/styles';
+import { DeprecatedThemeOptions } from '@mui/material';
+import { ThemeContext, ThemeSetter } from './ThemeContext';
+
+export const ThemeProvider = ({
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:12]: Please add JSDoc for usage

File: packages/ra-core/src/core/CoreAdmin.tsx [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { ComponentType } from 'react';
 
 import CoreAdminContext from './CoreAdminContext';
 import CoreAdminUI from './CoreAdminUI';
-import { AdminProps } from '../types';
+import { CoreAdminProps } from '../types';
```
Comment (fzaninotto) [inline packages/ra-core/src/core/CoreAdmin.tsx:6]: I think we should colocate this type with the component

File: packages/ra-ui-materialui/src/layout/Theme/ThemeContext.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { createContext } from 'react';
+import { DeprecatedThemeOptions } from '@mui/material';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Theme/ThemeContext.ts:2]: Frightening. Isn't there a better alternative?

File: packages/react-admin/package.json [hunk 5]
```diff
@@ -36,14 +36,12 @@
         "@emotion/styled": "^11.3.0",
         "@mui/icons-material": "^5.0.1",
         "@mui/material": "^5.0.2",
-        "final-form": "^4.20.2",
+        "final-form": "^4.20.4",
```
Comment (fzaninotto) [inline packages/react-admin/package.json:39]: seems unrelated to this PR

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Much cleaner like that!

But your PR lacks a section in the upgrade guide, and a section in the Theming doc

File: packages/react-admin/package.json (hunk 5 above)
Comment (djhi) [reply packages/react-admin/package.json:39]: That's just a line that moved, most probably automatically

File: packages/ra-ui-materialui/src/layout/Theme/useSetTheme.ts (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/Theme/useSetTheme.ts]: Added

File: packages/ra-core/src/core/CoreAdmin.tsx (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/core/CoreAdmin.tsx]: Done but a lot of other types could be collocated as well

File: packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:19]: This is almost an internal component. Do you want an example that would be a copy of `react-admin` `Admin` component?

File: packages/react-admin/src/Admin.tsx [hunk 6]
```diff
@@ -125,32 +127,37 @@ const Admin = (props: AdminProps) => {
...
-                disableTelemetry={disableTelemetry}
-                menu={menu}
-                catchAll={catchAll}
-                theme={theme}
-                title={title}
-                loading={loading}
-                loginPage={loginPage}
-                logout={authProvider ? logoutButton : undefined}
-                ready={ready}
+        <ThemeProvider theme={theme}>
```
Comment (fzaninotto) [inline packages/react-admin/src/Admin.tsx:130]: I think the ThemeProvider should be in the AdminContext. It would facilitate our tests, as we'd only have to wrap mui components with AdminContext (just like we wrap core components with CoreAdminContext).

File: packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx (hunk 2 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:19]: No, that's fine.

File: packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        </ThemeContext.Provider>
+    );
+};
+
+export interface ThemeProviderProps {
+    children: ReactNode;
+    theme: ThemeOptions;
+}
+
+const useTheme = (themeOverride?: ThemeOptions): [Theme, ThemeSetter] => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:37]: This one should be in its own file, with an example usage in jsDoc

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Still missing docs and upgrade instructions

File: packages/react-admin/src/Admin.tsx (hunk 6 above)
Comment (djhi) [reply packages/react-admin/src/Admin.tsx:130]: Sure

File: packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx (hunk 7 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:37]: We don't even export it. It's only used internally by the `ThemeProvider`

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx:37]: I think I'm confused because you didn't document how to read and set the theme ;)

File: UPGRADE.md [hunk 8]
```diff
@@ -1209,6 +1209,35 @@ Besides, some props supported by the previous implementation aren't anymore:
 
 `useGetMainList` was a modified version of `useGetList` designed to keep previous data on screen upon navigation. As [this is now supported natively by react-query](https://react-query.tanstack.com/guides/paginated-queries#better-paginated-queries-with-keeppreviousdata), this hook is no longer necessary and has been removed. Use `useGetList()` instead.
 
+## The MUI `<ThemeProvider>` is not set by `<Layout>` anymore
+
+This shouldn't impact your code unless you had a completely custom `<Layout>` component. If you do but still uses the default `<Admin>` component, you can safely remove the `ThemeProvider` from your `Layout`:
```
Comment (fzaninotto) [inline UPGRADE.md:1214]: This is a bit misleading, so developers may wonder if they need to manually add the ThemeProvider themselves. You should explain that ThemeProvider is set by AdminContext rather than Layout

File: docs/Theming.md [hunk 9]
```diff
@@ -332,6 +332,34 @@ const App = () => (
 );
 ```
 
+## Changing the theme programmatically
```
Comment (fzaninotto) [inline docs/Theming.md:335]: use Title Case

**Review by fzaninotto — CHANGES_REQUESTED** [review]
tests fail

File: UPGRADE.md (hunk 8 above)
Comment (djhi) [reply UPGRADE.md:1216]: Not sure why github shows the old version but I added a sentence before this one as requested
